src/Compilers/Server/VBCSCompiler/ClientConnectionHandler.cs

```diff
@@ -0,0 +1,174 @@
+            {
+                using var clientConnection = await clientConnectionTask.ConfigureAwait(false);
+                var request = await clientConnection.ReadBuildRequestAsync(cancellationToken).ConfigureAwait(false);
+
+                if (request.ProtocolVersion != BuildProtocolConstants.ProtocolVersion)
+                {
+                    var response = new MismatchedVersionBuildResponse();
+                    await clientConnection.WriteBuildResponseAsync(response, cancellationToken).ConfigureAwait(false);
+                    return CompletionData.RequestCompleted;
+                }
+
+                if (!string.Equals(request.CompilerHash, BuildProtocolConstants.GetCommitHash(), StringComparison.OrdinalIgnoreCase))
+                {
+                    var response = new IncorrectHashBuildResponse();
+                    await clientConnection.WriteBuildResponseAsync(response, cancellationToken).ConfigureAwait(false);
+                    return CompletionData.RequestCompleted;
+                }
+
+                if (request.Arguments.Count == 1 && request.Arguments[0].ArgumentId == BuildProtocolConstants.ArgumentId.Shutdown)
+                {
+                    var id = Process.GetCurrentProcess().Id;
+                    var response = new ShutdownBuildResponse(id);
+                    await clientConnection.WriteBuildResponseAsync(response, cancellationToken).ConfigureAwait(false);
+                    return new CompletionData(CompletionReason.RequestCompleted, shutdownRequested: true);
+                }
+
+                if (!allowCompilationRequests)
+                {
+                    var response = new RejectedBuildResponse("Compilation not allowed at this time");
+                    await clientConnection.WriteBuildResponseAsync(response, cancellationToken).ConfigureAwait(false);
+                    return CompletionData.RequestCompleted;
+                }
+
+                return await ProcessCompliationRequestAsync(clientConnection, request, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private async Task<CompletionData> ProcessCompliationRequestAsync(IClientConnection clientConnection, BuildRequest request, CancellationToken cancellationToken)
+        {
+            // Need to wait for the compilation and client disconnection in parallel. If the client
+            // suddenly disconnects we need to cancel the compilation that is occuring. It could be the 
+            // client hit Ctrl-C due to a run away analyzer.
+            var buildCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            var compilationTask = ProcessCompilationRequestCore(CompilerServerHost, request, buildCancellationTokenSource.Token);
+            await Task.WhenAny(compilationTask, clientConnection.DisconnectTask).ConfigureAwait(false);
+
+            try
+            {
+                if (compilationTask.IsCompleted)
```

Comment:
Couldn't it be that client disconnected _and_ the compilation completed? It seems that we're just writing to the client connection in that case, is that ok?

Reply:
Yes that can happen but in that case the write to the client of the response will fail.